File: promptc/resolver.py
"""Import resolution for prompt files."""

from typing import List, Set, Dict
from pathlib import Path
from .library import LibraryManager


class ImportResolver:
    """Resolves imports recursively with cycle detection and exclusions."""

    def __init__(self, library_manager: LibraryManager):
        """
        Initialize resolver.

        Args:
            library_manager: Library manager for resolving imports
        """
        self.library_manager = library_manager
        self._visited: Set[str] = set()
        self._exclusions: Set[str] = set()
        self._resolved_content: Dict[str, str] = {}

    def resolve(self, imports: List[str]) -> str:
        """
        Resolve all imports recursively.

        Args:
            imports: List of import names (may include ! prefix for exclusions)

        Returns:
            Combined content from all resolved imports
        """
        self._visited.clear()
        self._exclusions.clear()
        self._resolved_content.clear()

        # First pass: collect exclusions
        for import_name in imports:
            if import_name.startswith("!"):
                self._exclusions.add(import_name[1:])

        # Second pass: resolve imports
        content_parts = []
        for import_name in imports:
            if not import_name.startswith("!"):
                resolved = self._resolve_recursive(import_name)
                if resolved:
                    content_parts.append(resolved)

        return "\n\n".join(content_parts)

    def _resolve_recursive(self, import_name: str) -> str:
        """
        Recursively resolve a single import.

        Args:
            import_name: Import name to resolve

        Returns:
            Combined content from this import and its dependencies
        """
        # Check if excluded
        if import_name in self._exclusions:
            return ""

        # Check if already visited (cycle detection)
        if import_name in self._visited:
            return ""

        # Check if already resolved
        if import_name in self._resolved_content:
            return self._resolved_content[import_name]

        self._visited.add(import_name)

        # Load the content
        try:
            content = self.library_manager.load(import_name)
        except FileNotFoundError as e:
            raise FileNotFoundError(f"Failed to resolve import '{import_name}': {e}")

        # Store and return
        self._resolved_content[import_name] = content
        return content

    def get_resolution_order(self) -> List[str]:
        """
        Get the order in which imports were resolved.

        Returns:
            List of import names in resolution order
        """
        return list(self._visited)

File: internal/resolver/resolver.go
package resolver

import (
	"fmt"
	"strings"

	"github.com/Geogboe/promptc/internal/library"
)

// Resolver resolves imports recursively with cycle detection
type Resolver struct {
	libraryManager *library.Manager
	visited        map[string]bool
	exclusions     map[string]bool
	resolved       map[string]string
}

// NewResolver creates a new import resolver
func NewResolver(libraryManager *library.Manager) *Resolver {
	return &Resolver{
		libraryManager: libraryManager,
		visited:        make(map[string]bool),
		exclusions:     make(map[string]bool),
		resolved:       make(map[string]string),
	}
}

// Resolve resolves all imports recursively
func (r *Resolver) Resolve(imports []string) (string, error) {
	// Reset state
	r.visited = make(map[string]bool)
	r.exclusions = make(map[string]bool)
	r.resolved = make(map[string]string)

	// First pass: collect exclusions
	for _, imp := range imports {
		if strings.HasPrefix(imp, "!") {
			r.exclusions[imp[1:]] = true
		}
	}

	// Second pass: resolve imports
	var contentParts []string
	for _, imp := range imports {
		if !strings.HasPrefix(imp, "!") {
			content, err := r.resolveRecursive(imp)
			if err != nil {
				return "", err
			}
			if content != "" {
				contentParts = append(contentParts, content)
			}
		}
	}

	return strings.Join(contentParts, "\n\n"), nil
}

func (r *Resolver) resolveRecursive(importName string) (string, error) {
	// Check if excluded
	if r.exclusions[importName] {
		return "", nil
	}

	// Check if already visited (cycle detection)
	if r.visited[importName] {
		return "", nil
	}

	// Check if already resolved
	if content, ok := r.resolved[importName]; ok {
		return content, nil
	}

	r.visited[importName] = true

	// Load the content
	content, err := r.libraryManager.Resolve(importName)
	if err != nil {
		return "", fmt.Errorf("failed to resolve import '%s': %w", importName, err)
	}

	// Store and return
	r.resolved[importName] = content
	return content, nil
}

// GetResolutionOrder returns the order in which imports were resolved
func (r *Resolver) GetResolutionOrder() []string {
	order := make([]string, 0, len(r.visited))
	for imp := range r.visited {
		order = append(order, imp)
	}
	return order
}
